display/src/main/scala/uk/ac/wellcome/platform/archive/display/DisplayLocation.scala

```diff
@@ -12,19 +15,37 @@ case class DisplayLocation(
   @JsonKey("type") ontologyType: String = "Location"
 ) {
   def toSourceLocation: SourceLocation =
-    SourceLocation(
-      provider = provider.toStorageProvider,
-      location = ObjectLocation(bucket, path)
-    )
+    provider.toStorageProvider match {
+      case AmazonS3StorageProvider =>
+        S3SourceLocation(
+          prefix = S3ObjectLocationPrefix(bucket = bucket, keyPrefix = path)
+        )
+
+      case AzureBlobStorageProvider =>
+        AzureBlobSourceLocation(
+          prefix =
+            AzureBlobItemLocationPrefix(container = bucket, namePrefix = path)
+        )
+    }
 }
 
 object DisplayLocation {
   def apply(location: SourceLocation): DisplayLocation =
-    DisplayLocation(
-      provider = DisplayProvider(location.provider),
-      bucket = location.location.namespace,
-      path = location.location.path
-    )
+    location match {
+      case S3SourceLocation(prefix) =>
+        DisplayLocation(
+          provider = DisplayProvider(location.provider),
+          bucket = prefix.bucket,
+          path = prefix.keyPrefix
+        )
+
+      case AzureBlobSourceLocation(prefix) =>
```

Comment:
I think we should check whether this will affect clients of the storage APIs (i.e. DDS).

Reply:
I don't think so – they’ll receive the same block of JSON as before, but whether you construct it from bucket/key or container/name depends on the provider. If I've done it right, this should be an entirely internal change that clients don't notice.

Reply:
It occurs to me we also don't need this change, now I added namespace/path methods to the Location model.